### PR TITLE
Serial Communication

### PR DESCRIPTION
**Pull request by TekuConcept**
Secures the cross communication between Arduino a Bone with fast primitive data serialization. (Also even allows for the serialization of structs and other non-primitive types - other than pointers.)

### CODE REVIEW

File: ArduinoController/ArduinoController.ino [hunk 1]
```diff
@@ -10,43 +10,41 @@
...
-const uint8_t KILLPIN = 50; // Needs to remain global (therefore outside of JSON) for lambda interrupt
-const uint8_t KILL_ADDR = 10;
 const uint32_t CONTROLLER_CNT = 13u;
 class IController* controllers[CONTROLLER_CNT];
 
 void setup() {
   Serial.begin(115200);
-  Serial.print('R');
-  DynamicJsonBuffer buffer;
-  auto json = buffer.parse(PinJSON::json);
```
Comment (nfcopier) [inline ArduinoController/ArduinoController.ino]: Why did you remove the JSON? One of our goals is to be able to reconfigure the Arduino pin map live from the Beaglebone by sending new JSON. This moves in the opposite direction from that.

File: BoneCentral/Peripherals/Peripherals.SerialAdaptors/include/SerialThruster.h [hunk 2]
```diff
@@ -7,22 +7,14 @@
 
 #include <IThruster.h>
 #include "Serial.h"
-
-enum ThrusterIndex {
```
Comment (TekuConcept) [inline BoneCentral/Peripherals/Peripherals.SerialAdaptors/include/SerialThruster.h]: That is a good question! Why did I do that?

File: BoneCentral/Peripherals/Peripherals.SerialAdaptors/src/Serial.cpp [hunk 3]
```diff
@@ -1,63 +1,191 @@
...
+    topts.c_oflag &= ~OPOST;
+    topts.c_cc[VMIN] = 1;
+    topts.c_cc[VTIME] = 0;
+    if(tcsetattr(fd, TCSANOW, &topts)) {
+        LOG("Failed to set terminal ios options for file descriptor.\n");
+        throw 1;
+    }
+    if(tcflush(fd, TCIFLUSH)) {
+        LOG("Failed to flush file descriptor.\n");
+        throw 1;
```
Comment (TekuConcept) [inline BoneCentral/Peripherals/Peripherals.SerialAdaptors/src/Serial.cpp]: To be honest, this was a copy-past from Daniel Tonk's code. Other than the `cfsetispeed` and `cfsetospeed`, I'm not exactly sure how else I can break it up. All the c_Xflag (configuration flags) options are the same options passed in the original system call.

File: ArduinoController/ArduinoController.ino (hunk 1 above)
Comment (TekuConcept) [inline ArduinoController/ArduinoController.ino]: I knew this day would come when I would need to explain my self... In fact, I prepared by writing a several page exploratory research paper of the pros and cons of using JSON with microcontrollers and serial communication.

In summary, serial UART communication is not near as fast as traditional USB2.0 as one might think. To demonstrate, some JSON strings could take up to 250 ms to transfer with serial UART when it would only take 2.9 ms via USB2.0. (That's 86 times slower)

While this move appears to be a step backward, I actually took this goal into consideration when writing the serialization subroutines. The two base functions `readData` and write `writeData` don't just serialize primitive types.

As the test, `Test_StructDataType`, demonstrates, user-defined structs may also be serialized/deserialized between devices. With this, a simple pin configuration struct can be created for both sides (which is required anyway) and then pins can be configured in one blazing fast call.

File: BoneCentral/Peripherals/Peripherals.SerialAdaptors/src/Serial.cpp [hunk 4]
```diff
@@ -1,63 +1,191 @@
...
+    if(fd==0) return "";
+    std::stringstream ss;
+    char c[1];
+    
+    do{
+        read(fd, c, 1);
+        ss << c[0];
+    }while(c[0] != '\0');
+    return ss.str();
+#endif
```
Comment (TekuConcept) [inline BoneCentral/Peripherals/Peripherals.SerialAdaptors/src/Serial.cpp]: This morning I read up on professional style guides (Google C++ Style Guide) and they are in agreement with you.

https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros

I will try to find alternative approaches that won't litter the code with preprocessor directives but also won't double the code base either.

I will ommit any UNO related code since our project deals solely with the Due.

The problem with implementing two classes (Due and Debug) is that the primary purpose of the debug mode was to actively see and "debug" the flow of information in the main code.

Serial here, of course, probably could have its own "dummy" object, but a preprocessor directive (or an update to cmake) will still be required.
